File: iosApp/iosAppTests/Pages/Map/MapHttpInterceptorTests.swift
//
//  MapHttpInterceptorTests.swift
//  iosAppTests
//
//  Created by Brady, Kayla on 7/24/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

@testable import iosApp
import XCTest
@_spi(Experimental) import MapboxMaps

final class MapHttpInterceptorTests: XCTestCase {
    override func setUp() {
        executionTimeAllowance = 60
    }

    let fakeRequest: HttpRequest = .init(url: "fake_url",
                                         headers: [:],
                                         sdkInformation: .init(name: "name", version: "v", packageName: nil),
                                         body: nil)

    func testOnRequest() throws {
        var requestUnchanged = false
        let interceptor = MapHttpInterceptor(updateLastErrorTimestamp: {})
        interceptor.onRequest(for: fakeRequest) { requstOrResponse in
            if requstOrResponse.getHttpRequest() == self.fakeRequest {
                requestUnchanged = true
            }
        }

        XCTAssertTrue(requestUnchanged)
    }

    func testOnResponse200DoesNothing() throws {
        let fake200Response: HttpResponse = .init(identifier: 1, request: fakeRequest,
                                                  result: .success(.init(headers: [:], code: 200, data: .init())))
        var responseUnchanged = false
        var updateLastErrorCalled = false

        let interceptor = MapHttpInterceptor(updateLastErrorTimestamp: { updateLastErrorCalled = true })
        interceptor.onResponse(for: fake200Response, continuation: { response in
            if response == fake200Response {
                responseUnchanged = true
            }
        })

        XCTAssertTrue(responseUnchanged)
        XCTAssertFalse(updateLastErrorCalled)
    }

    func testOnResponse401UpdatesLastError() throws {
        let fake401Response: HttpResponse = .init(identifier: 1, request: fakeRequest,
                                                  result: .success(.init(headers: [:], code: 401, data: .init())))
        var responseUnchanged = false
        var updateLastErrorCalled = false

        let interceptor = MapHttpInterceptor(updateLastErrorTimestamp: { updateLastErrorCalled = true })
        interceptor.onResponse(for: fake401Response, continuation: { response in
            if response == fake401Response {
                responseUnchanged = true
            }
        })

        XCTAssertTrue(responseUnchanged)
        XCTAssertTrue(updateLastErrorCalled)
    }

    func testOnRequestCancelledDoesNothing() throws {
        let fakeErrorResponse: HttpResponse = .init(identifier: 1, request: fakeRequest,
                                                    result: .failure(.init(type: .requestCancelled,
                                                                           message: "Cancelled")))
        var responseUnchanged = false
        var updateLastErrorCalled = false

        let interceptor = MapHttpInterceptor(updateLastErrorTimestamp: { updateLastErrorCalled = true })
        interceptor.onResponse(for: fakeErrorResponse, continuation: { response in
            if response == fakeErrorResponse {
                responseUnchanged = true
            }
        })

        XCTAssertTrue(responseUnchanged)
        XCTAssertFalse(updateLastErrorCalled)
    }
}

File: androidApp/src/androidTest/java/com/mbta/tid/mbta_app/android/map/MapHttpInterceptorTests.kt
package com.mbta.tid.mbta_app.android.map

import com.mapbox.bindgen.ExpectedFactory
import com.mapbox.common.HttpMethod
import com.mapbox.common.HttpRequest
import com.mapbox.common.HttpRequestError
import com.mapbox.common.HttpRequestErrorType
import com.mapbox.common.HttpResponse
import com.mapbox.common.HttpResponseData
import com.mapbox.common.SdkInformation
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest
import org.junit.Test

class MapHttpInterceptorTests {
    private val fakeRequest: HttpRequest =
        HttpRequest.Builder()
            .method(HttpMethod.GET)
            .url("fake_url")
            .headers(hashMapOf())
            .sdkInformation(SdkInformation("a", "b", "c"))
            .build()

    @Test
    fun testOnRequestUnchanged() = runTest {
        var requestUnchanged = false
        val interceptor = MapHttpInterceptor {}
        interceptor.onRequest(fakeRequest) { requstOrResponse ->
            if (requstOrResponse.getHttpRequest() == fakeRequest) {
                requestUnchanged = true
            }
        }

        assertTrue { requestUnchanged }
    }

    @Test
    fun testOnResponse200DoesNothing() = runTest {
        val fakeResponse: HttpResponse =
            HttpResponse(
                1,
                fakeRequest,
                ExpectedFactory.createValue(HttpResponseData(hashMapOf(), 200, byteArrayOf()))
            )

        var responseUnchanged = false
        var updateLastErrorCalled = false
        val interceptor = MapHttpInterceptor { updateLastErrorCalled = true }
        interceptor.onResponse(fakeResponse) { requstOrResponse ->
            if (requstOrResponse.request == fakeRequest) {
                responseUnchanged = true
            }
        }

        assertTrue { responseUnchanged }
        assertFalse { updateLastErrorCalled }
    }

    @Test
    fun testOn401UpdatesLastError() = runTest {
        val fakeResponse =
            HttpResponse(
                1,
                fakeRequest,
                ExpectedFactory.createValue(HttpResponseData(hashMapOf(), 401, byteArrayOf()))
            )

        var responseUnchanged = false
        var updateLastErrorCalled = false
        val interceptor = MapHttpInterceptor { updateLastErrorCalled = true }
        interceptor.onResponse(fakeResponse) { requstOrResponse ->
            if (requstOrResponse.request == fakeRequest) {
                responseUnchanged = true
            }
        }

        assertTrue { responseUnchanged }
        assertTrue { updateLastErrorCalled }
    }

    @Test
    fun testOnRequsetCancelledDoesNothing() = runTest {
        val fakeResponse =
            HttpResponse(
                1,
                fakeRequest,
                ExpectedFactory.createError(
                    HttpRequestError(HttpRequestErrorType.REQUEST_CANCELLED, "cancelled")
                )
            )

        var responseUnchanged = false
        var updateLastErrorCalled = false
        val interceptor = MapHttpInterceptor { updateLastErrorCalled = true }
        interceptor.onResponse(fakeResponse) { requstOrResponse ->
            if (requstOrResponse.request == fakeRequest) {
                responseUnchanged = true
            }
        }

        assertTrue { responseUnchanged }
        assertFalse { updateLastErrorCalled }
    }
}
